_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,791 @@
+    url: String,
+    alias: String,
+    data: Vec<WebsiteStats>,
+}
+```
+
+To start working on our backend, we can create an initial route to add a URL to monitor. You may have noticed earlier we added the `Validate` derive trait. This allows us to validate the form data using preset rules and automatically return an error if the validation fails. In this case, we used `#[validate(url)]` - so if the string isn’t in a URL format it will automatically break:
+
+```rust
+// main.rs
+async fn create_website(
+    State(state): State<AppState>,
+    Form(new_website): Form<Website>,
+) -> Result<impl AxumIntoResponse, impl AxumIntoResponse> {
+    if new_website.validate().is_err() {
+        return Err((
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "Validation error: is your website a reachable URL?",
+        ));
+    }
+
+    sqlx::query("INSERT INTO websites (url, alias) VALUES ($1, $2)")
+        .bind(new_website.url)
+        .bind(new_website.alias)
+        .execute(&state.db)
+        .await
+        .unwrap();
+
+    Ok(Redirect::to("/"))
+```
+
+Now let’s write a route to grab all the websites we’re monitoring, as well as get a quick report on what the uptime was like in the last recorded 24 hours (filling in any gaps where required) for each website. 
+
+Like before with the monitoring tasks, we need to grab a list of all of the websites we’re currently tracking (except we will assume there are results there. If not, `askama` will handle it for us by automatically not rendering any records):
+
+```rust
+async fn get_websites(State(state): State<AppState>) -> impl AskamaIntoResponse {
+    let websites = sqlx::query_as::<_, Website>("SELECT url, alias FROM websites")
+        .fetch_all(&state.db)
+        .await
+        .unwrap();
+    // .. rest of your code
+}
+```
+
+Once this is done, we’ll then need to make a new Vector. This will hold all the website URLs (and respective aliases) as well as a list of timestamps with the uptime percentage over the last 24 hours, calculated per hour from how many HTTP requests returned with `200 OK`. We then need to check if there’s any gaps and fill them in with a `None`. This lets the person viewing the data know that no data was recorded at the time (for example if we’re either developing locally, or if the service itself had an outage and was unable to record data). 
+
+```rust
+// main.rs
+let mut logs = Vec::new();
+
+for website in websites {
+    let mut data = sqlx
+        ::query_as::<_, WebsiteStats>(
+            r#"
+            SELECT date_trunc('hour', created_at) as time,
+            CAST(COUNT(case when status = 200 then 1 end) * 100 / COUNT(*) AS int2) as uptime_pct
+            FROM logs WHERE website_alias = $1
+            group by time
+            order by time asc
+            limit 24
+            "#
+        )
+        .bind(&website.alias)
+        .fetch_all(&state.db).await
+        .unwrap();
+
+    if data.len() < 24 {
+        for i in 1..24 {
```

Comment:
Suppose this service was down for 3 hours about 5 hours ago, then what will this do? Keep in mind that your query might have returned the 24 rows if all the data for one day ago is complete.